### PR TITLE
Fix over-eager file matching in skip files and `--file` option

### PR DESCRIPTION
**Pull request by furtib**
Why:
In the current implementation of the skipfiles, every file declared will match any other file that contains the declared file name at the beginning of its name. An example would be: `simple.c` and `simple.cpp`. Specifying `simple.c` with the `--flag` option would result in a regex: `/path/to/simple.c.*`

What:
Removed adding '*' to the end of any regex given through `--file` or the skipfile.

Notes:
According to the [documentation](https://codechecker.readthedocs.io/en/latest/analyzer/user_guide/#skip-file), this change is breaking, but I could see some users declaring directories in their skip file without placing a '*' at the end.

Fixes: #4664

### CODE REVIEW

**Comment by furtib** [comment]
I have added a clause to add the star to entries that are evidently folders.
Users will use the '/' to point to folders even on Windows.
With this, I do respect that no filename *may* end with the '/' character.
Filenames can't contain the '/' character on [Windows](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#:~:text=Use%20any%20character%20in%20the%20current%20code%20page%20for%20a%20name%2C%20including%20Unicode%20characters%20and%20characters%20in%20the%20extended%20character%20set%20(128%E2%80%93255)%2C%20except%20for%20the%20following%3A). On linux I think common sense might stop users from ending their filenames with a '/'.

@barnabasdomozi, what are your thoughts?

**Comment by furtib** [comment]
I cannot be sure if the path you gave is for a file named build or for a folder, so I do not handle that case.

During a conversation with Botond, I got the suggestion to instead of adding a '*' after a '/' we could add an optional '/.*' to the end of our regex. This would look something like this: `(?:\/.*)`?.

This would solve the issue of `simple.c` and `simple.cc` while also preserving the workings for a directory path defined like: `/workspace/myproject/build`

File: codechecker_common/skiplist_handler.py [hunk 1]
```diff
@@ -52,8 +52,15 @@ def __gen_regex(self, skip_lines):
...
+            # fnmatch places a '\Z' at the end, this  is equivalent to '$'
+            # (end of input) so it must be removed.
+            # Optionally we match the user given path with a '/.*' ending.
+            # This, if the given input is a folder will
+            # resolve all files contained within.
+            # Note: normalization removes '/' from the end, see:
+            # https://docs.python.org/3/library/os.path.html#os.path.normpath
             rexpr = re.compile(
-                fnmatch.translate(norm_skip_path + '*'))
+                fnmatch.translate(norm_skip_path)[:-2] + r"(?:/.*)?\Z")
```
Comment (barnabasdomozi) [inline codechecker_common/skiplist_handler.py]: If `fnmatch.translate()` function's implementation changes, e.g. it no longer puts `\Z` in the end, then `[:-2]` breaks the regex. So instead of `[:-2]` consider using `rstrip()`

Also, is there a reason we are using `\Z` in the end of the regex and not `$`?

**Review by barnabasdomozi — CHANGES_REQUESTED** [review]
Great work!
See my comment about a minor issue.

We should also investigate which slash type is used on Windows.

File: codechecker_common/skiplist_handler.py (hunk 1 above)
Comment (furtib) [reply codechecker_common/skiplist_handler.py]: I wrote a simple check to see if \Z is there, before removing it.
rstrip(r"\Z") would have stripped away any path ending in z.
e.g: "abcz\Z" -> "abc"

**Comment by furtib** [comment]
I have added a check to the fnmatch. translate output, so as not to depend so much on its implementation.
I also tried to support Windows systems. `os.path.normalize("/")` will return `"\\"` (an escaped backslash). This way, we will check for an optional `\\,*` on Windows

File: analyzer/tests/functional/skip/test_skip.py [hunk 2]
```diff
@@ -170,13 +171,66 @@ def test_skip(self):
                             "Reports from headers should be kept"
                             " if the header is not on the skiplist")
 
+    def test_skip_directory(self):
+        """Test wether directories in skipfile entries works correctly"""
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]: Typo: "wether" should be "whether"

File: analyzer/tests/functional/skip/test_skip.py [hunk 3]
```diff
@@ -482,11 +549,23 @@ def test_analyze_file_option(self):
                 any('skip_header.cpp' not in f for f in glob.glob(
                     os.path.join(self.report_dir, '*.plist'))))
 
+    def test_analyze_similar_file_option(self):
+        """
+        Test analyze command --file option for edge case,
+        where one filename is a prefix of an other.
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]: Typo: another

File: analyzer/tests/functional/skip/test_skip.py [hunk 4]
```diff
@@ -170,13 +171,66 @@ def test_skip(self):
...
+            skip_file.flush()
+            self.__log_and_analyze("skip_folder", ["--ignore", skip_file.name])
+
+            # Check if the folder is skipped
+            # There shouldn't be any report generated
+            report_dir_files = os.listdir(self.report_dir)
+            for f in report_dir_files:
+                self.assertFalse("skipme.cpp" in f)
+
+        with tempfile.NamedTemporaryFile(
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]: These 3 code blocks seem redundant and only differ in skipfile content.
Consider creating a loop for the cases `['-*skipme/*', '-*skipme/', '-*skipme']` or define a helper function.

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 5]
```diff
@@ -0,0 +1,14 @@
+OBJS = $(SRCS:.cpp=.o)
+
+CXXFLAGS = -Wno-all -Wno-extra -Wno-division-by-zero
+
+SRCS = simple.c \
+		simple.cc
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile]: We usually use the `.cpp` extension rather than `.cc` throughout the C++ test files.

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+.cpp.o:
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile]: Currently there are no `.cpp` files in this folder. There is also no rule for `.c` files.

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 7]
```diff
@@ -0,0 +1,14 @@
+OBJS = $(SRCS:.cpp=.o)
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile]: This Makefile looks incomplete.
`OBJS` is not defined for `.c` or `.cc` files. Also note there are no `.cpp` files in this directory.

File: analyzer/tests/functional/skip/test_skip.py [hunk 8]
```diff
@@ -170,13 +171,66 @@ def test_skip(self):
...
+            mode="w", suffix="skipfile", encoding="utf-8"
+        ) as skip_file:
+
+            # Skip the skipme folder, without any indication that its a folder
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]: Nit: it should be `it's` or `it is`. `its` is incorrect in this sentence and has a different meaning.

File: codechecker_common/skiplist_handler.py [hunk 9]
```diff
@@ -52,8 +52,18 @@ def __gen_regex(self, skip_lines):
...
+            # This, if the given input is a folder will
+            # resolve all files contained within.
+            # Note: normalization removes '/' from the end, see:
+            # https://docs.python.org/3/library/os.path.html#os.path.normpath
+            translated_glob = fnmatch.translate(norm_skip_path)
+            if translated_glob.endswith(r"\Z"):
+                translated_glob = translated_glob[:-2]
             rexpr = re.compile(
-                fnmatch.translate(norm_skip_path + '*'))
+                translated_glob + fr"(?:\{os.path.sep}.*)?$")
```
[the remark this answers is not in the document]
Comment (furtib) [reply codechecker_common/skiplist_handler.py:66]: There is a long comment before it explaining what it does, and its purpose.

File: analyzer/tests/functional/skip/test_skip.py [hunk 10]
```diff
@@ -482,11 +521,23 @@ def test_analyze_file_option(self):
...
+    def test_analyze_similar_file_option(self):
+        """
+        Test analyze command --file option for edge case,
+        where one filename is a prefix of another.
+        """
+        self.__log_and_analyze("similar", ["--file", "*/simple.c"])
+        print(glob.glob(
+            os.path.join(self.report_dir, '*.plist')))
+        self.assertFalse(
+            any('simple.cpp' not in f for f in glob.glob(
```
[the remark this answers is not in the document]
Comment (furtib) [reply analyzer/tests/functional/skip/test_skip.py]: You are right!

File: analyzer/tests/functional/skip/test_skip.py [hunk 11]
```diff
@@ -170,13 +171,38 @@ def test_skip(self):
...
+            ["-*skipme/*"],  # The way its in the docs
+            ["-*skipme/"],  # If the user specified that this is a directory
+            ["-*skipme"],  # If we can't even be sure that this is a directory
+        ]
+        for skip_file_content in skip_file_contents:
+            with tempfile.NamedTemporaryFile(
+                mode="w", suffix="skipfile", encoding="utf-8"
+            ) as skip_file:
+
+                # Skip the `skipme` folder, the way its in the documentation
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]: Please update the comment here as well.

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+CPP_OBJS = $(CPP_SRCS:.cpp=_cpp.o)
+
+OBJS =  $(C_OBJS) $(CPP_OBJS)
+
+CXXFLAGS = -Wno-all -Wno-extra -Wno-division-by-zero
+
+C_SRCS = simple.c
+CPP_SRCS = simple.cpp
+
+all: $(C_OBJS) $(CPP_OBJS)
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile]: Use `$(OBJS)` here.

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 13]
```diff
@@ -0,0 +1,20 @@
+C_OBJS = $(C_SRCS:.c=_c.o)
+CPP_OBJS = $(CPP_SRCS:.cpp=_cpp.o)
+
+OBJS =  $(C_OBJS) $(CPP_OBJS)
+
+CXXFLAGS = -Wno-all -Wno-extra -Wno-division-by-zero
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile]: Instead of overwriting `$(CXXFLAGS)`, consider appending these flags to its default value.

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+C_SRCS = simple.c
+CPP_SRCS = simple.cpp
+
+all: $(C_OBJS) $(CPP_OBJS)
+
+$(CPP_OBJS): %_cpp.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $<  -o $@
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile]:
```suggestion
	$(CXX) $(CXXFLAGS) -c $< -o $@
```
Superfluous space.

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+$(C_OBJS): %_c.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
+
+clean:
+	rm -rf *.o
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile]:
```suggestion
	rm -rf $(OBJS)
```
Don't delete unrelated object files.

File: analyzer/tests/functional/skip/test_files/skip_folder/Makefile [hunk 16]
```diff
@@ -0,0 +1,13 @@
+OBJS = $(SRCS:.cpp=.o)
+
+CXXFLAGS = -Wno-all -Wno-extra -Wno-division-by-zero
+
+SRCS = skipme/skipme.cpp
+
+.cpp.o:
+	$(CXX) $(CXXFLAGS) -c $<  -o $@
+
+all: $(OBJS)
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/skip_folder/Makefile:7]: Move `all` up to ensure that it's the first rule (which is executed when `make` is ran without any arguments). I'm not sure that the "magical" rule `.cpp.o` can be the default, but it's better to avoid confusion.

File: analyzer/tests/functional/skip/test_files/skip_folder/Makefile [hunk 17]
```diff
@@ -0,0 +1,13 @@
+OBJS = $(SRCS:.cpp=.o)
+
+CXXFLAGS = -Wno-all -Wno-extra -Wno-division-by-zero
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/skip_folder/Makefile]: As above, append these flags to the default `$(CXXFLAGS)`.

File: analyzer/tests/functional/skip/test_files/skip_folder/Makefile [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+SRCS = skipme/skipme.cpp
+
+.cpp.o:
+	$(CXX) $(CXXFLAGS) -c $<  -o $@
+
+all: $(OBJS)
+
+clean:
+	rm -rf skipme/*.o
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/skip_folder/Makefile]:
```suggestion
	rm -rf $(OBJS)
```

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+C_SRCS = simple.c
+CPP_SRCS = simple.cpp
+
+all: $(C_OBJS) $(CPP_OBJS)
+
+$(CPP_OBJS): %_cpp.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $<  -o $@
+
+$(C_OBJS): %_c.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile:18]: You modified `$(CXXFLAGS)`, but you are using the default `$(CFLAGS)`. Is this your intention? (It is a reasonable default, if you don't need any customization.)

File: analyzer/tests/functional/skip/test_files/skip_folder/Makefile [hunk 20]
```diff
@@ -0,0 +1,13 @@
+OBJS = $(SRCS:.cpp=.o)
+
+CXXFLAGS = -Wno-all -Wno-extra -Wno-division-by-zero
+
+SRCS = skipme/skipme.cpp
+
+.cpp.o:
+	$(CXX) $(CXXFLAGS) -c $<  -o $@
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/skip_folder/Makefile]:
```suggestion
	$(CXX) $(CXXFLAGS) -c $< -o $@
```
[nit] Remove the extra space.

File: analyzer/tests/functional/skip/test_skip.py [hunk 21]
```diff
@@ -170,13 +171,38 @@ def test_skip(self):
...
+                skip_file.write('\n'.join(skip_file_content))
+                skip_file.flush()
+                self.__log_and_analyze(
+                    "skip_folder", ["--ignore", skip_file.name]
+                )
+
+                # Check if the folder is skipped
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]:
```suggestion
                # Check if the folder `skipme` is skipped
```

File: analyzer/tests/functional/skip/test_skip.py [hunk 22]
```diff
@@ -170,13 +171,38 @@ def test_skip(self):
...
+                # There shouldn't be any report generated
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]:
```suggestion
                # There shouldn't be any report generated for `skipme/skipme.cpp`
```

File: analyzer/tests/functional/skip/test_skip.py [hunk 23]
```diff
@@ -170,13 +171,39 @@ def test_skip(self):
...
+            ["-*skipme/*"],  # The way its in the docs
+            ["-*skipme/"],  # If the user specified that this is a directory
+            ["-*skipme"],  # If we can't even be sure that this is a directory
+        ]
+        for skip_file_content in skip_file_contents:
+            with tempfile.NamedTemporaryFile(
+                mode="w", suffix="skipfile", encoding="utf-8"
+            ) as skip_file:
+
+                # Skip the `skipme` folder, the way it is in the documentation
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]: `"the way it is in the documentation"`

This comment seems outdated since the code was modified to a loop which has 3 different cases.
```suggestion
                # Skip the `skipme` folder
```

File: analyzer/tests/functional/skip/test_skip.py [hunk 24]
```diff
@@ -170,13 +171,39 @@ def test_skip(self):
                             "Reports from headers should be kept"
                             " if the header is not on the skiplist")
 
+    def test_skip_directory(self):
+        """Test whether directories in skipfile entries works correctly"""
+        skip_file_contents: list[list[str]] = [
+            ["-*skipme/*"],  # The way its in the docs
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_skip.py]:
```suggestion
            ["-*skipme/*"],  # The way it is in the documentation
```

File: analyzer/tests/functional/skip/test_files/similar/Makefile [hunk 25]
```diff
@@ -0,0 +1,21 @@
+C_OBJS = $(C_SRCS:.c=_c.o)
+CPP_OBJS = $(CPP_SRCS:.cpp=_cpp.o)
+
+OBJS =  $(C_OBJS) $(CPP_OBJS)
```
Comment (barnabasdomozi) [inline analyzer/tests/functional/skip/test_files/similar/Makefile]:
```suggestion
OBJS = $(C_OBJS) $(CPP_OBJS)
```

[nit] Remove extra space.
